### PR TITLE
MNT: Refactor tuner forward methods for simplicity

### PR DESCRIPTION
**Pull request by BenjaminBossan**
This is to be consistent with changes in #794

## Description

The `forward` methods of several tuner layers were partly unnecessarily nested, which made them harder to read and which can conceal bugs (as in #794). Therefore, these methods have been refactored to (hopefully) be as readable as possible. Moreover, the different methods are now coded as similar as possible between the different implementations.

On top of those changes, I made the following adjustments:

- added some type hints to the methods of the layers
- removed a comment about code being copied which I think is not necessary
- for the lora embedding layer, we sometimes used `F.embedding(...)` and sometimes `nn.Embedding.forward(self, ...)` -- now, we consistently use only `F.embedding(...)`
- for IA³ `Linear8bitLT`, apply dtype conversion regardless of whether `self.is_feedforward` is `True` or `False` (as discussed internally)
- for the definition of lora `Linear4bit`, we (implicitly) checked if bnb is available and if bnb 4bit is available, but it is enough to check the latter, as it calls the former internally -- now we only check the latter, saving us 1 level of indentation (note that the github diff view of that class is very messy, in fact I only unindented it and changed `forward`, the rest is identical!)
- a few times, `ModuleDict.update(<dict>)` is called when `<dict>` has only 1 single item -- I simplified this code to just assign that item, which is more readable and consistent with other nearby code
- in the LoRA `Embedding.forward` method, there was no check if the `active_adapter` is in the lora module dict keys or not, unlike the other layer types -- now I added the check
- the error message when no LoRA layer could be replaced was outdated, only mentioning `Linear` and `Conv1D`, I added the other supported layer types

## Comments

In general, I ensured that the changes I made in the forward methods result in the exact same outcome as previously. Please ensure that I achieved this while reviewing this PR. This is probably easier with a side-by-side diff view.

While working on this, I found that a lot of code is essentially duplicated, but I think we can accept that if we want to avoid adding too many layers of abstraction/indirection.

Moreover, I discovered multiple potential inconsistencies between the implementations:

- sometimes, dtypes are converted, sometimes not
- dtypes are not always converted on the same intermediate result
- sometimes, `result.clone()` is called, sometimes not

I have added a `# FIXME` comment each time I spotted those differences. Before merging, all those comments should be addressed. This either means that we fix the inconsistencies, or, if they are intended, we add a comment to explain why they are intended.

### CODE REVIEW

File: src/peft/tuners/lora.py [hunk 1]
```diff
@@ -1145,91 +1131,98 @@ def __init__(
...
-                            self.lora_B[self.active_adapter](
-                                self.lora_A[self.active_adapter](self.lora_dropout[self.active_adapter](x))
-                            )
-                            * self.scaling[self.active_adapter]
-                        )
-                    result += output
                 return result
 
+            # FIXME: is it correct that we clone here but not for int8?
+            result = result.clone()
```
Comment (younesbelkada) [inline src/peft/tuners/lora.py:1208]: I see now, indeed this looks quite strange, I don't think we should clone here, I also realized we don't have 4bit training tests, I can maybe add a test with 4bit training and test if this works as expected

File: src/peft/tuners/lora.py [hunk 2]
```diff
@@ -1252,28 +1245,34 @@ def __init__(
...
-                output = (
-                    self.lora_B[self.active_adapter](
-                        self.lora_A[self.active_adapter](self.lora_dropout[self.active_adapter](x))
-                    )
-                    * self.scaling[self.active_adapter]
-                )
-            result += output
+
+        # FIXME: is it correct that we clone here but not for int8?
+        result = result.clone()
```
Comment (younesbelkada) [inline src/peft/tuners/lora.py:1259]: Same comment as above, I don't think we should clone here, we do have GPTQ tests, maybe we can run the test and check if it works with and without `clone`?

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this refactor, this looks super clean to me, I added two comments, I think we should remove the `clone()` calls, let's just make sure QLoRA training works after that, for that I can add 4bit training tests in a separate PR.

File: src/peft/tuners/lora.py (hunk 1 above)
Comment (younesbelkada) [reply src/peft/tuners/lora.py:1208]: Made: https://github.com/huggingface/peft/pull/834

**Review by younesbelkada — APPROVED** [review]
Thanks a lot ! As discussed internally let's not merge it before the next release

**Review by pacman100 — APPROVED** [review]
Thank you @BenjaminBossan for great work on refactoring, great improvement in readability, LGTM! 🤗

**Pull request merged**